File: frontend/src/api/competitions.ts
import client from "./client";
import type { Competition, CompetitionSummary, MyCompetitionSummary } from "../types";

export async function createCompetition(req: {
  name: string; club_id: number | null; format: string; courts: number; player_ids?: number[]; scheduled_at?: string;
  open_signup?: boolean; is_public?: boolean; max_players?: number; signup_deadline?: string;
}): Promise<Competition> {
  const { data } = await client.post("/competitions", req);
  return data;
}

export async function getCompetition(id: number): Promise<Competition> {
  const { data } = await client.get(`/competitions/${id}`);
  return data;
}

export async function listClubCompetitions(clubId: number): Promise<CompetitionSummary[]> {
  const { data } = await client.get(`/clubs/${clubId}/competitions`);
  return data;
}

export async function listOpenCompetitions(q = ""): Promise<CompetitionSummary[]> {
  const { data } = await client.get("/competitions/open", { params: { q: q || undefined } });
  return data;
}

export async function listMyCompetitions(): Promise<MyCompetitionSummary[]> {
  const { data } = await client.get("/competitions/me");
  return data;
}

export async function joinCompetition(id: number): Promise<Competition> {
  const { data } = await client.post(`/competitions/${id}/join`);
  return data;
}

export async function leaveCompetition(id: number): Promise<Competition> {
  const { data } = await client.delete(`/competitions/${id}/join`);
  return data;
}

export async function startCompetition(id: number): Promise<Competition> {
  const { data } = await client.patch(`/competitions/${id}/start`);
  return data;
}

export async function recordScore(matchId: number, scoreA: number, scoreB: number) {
  const { data } = await client.post(`/matches/${matchId}/score`, { score_a: scoreA, score_b: scoreB });
  return data;
}

export async function finishCompetition(id: number): Promise<Competition> {
  const { data } = await client.patch(`/competitions/${id}/finish`);
  return data;
}

File: backend/app/routers/competitions.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.deps import get_current_user
from app.models.user import User
from app.schemas.competition import (
    CompetitionResponse, CompetitionSummary, CreateCompetitionRequest, MatchResponse, MyCompetitionSummary, ScoreRequest,
)
from app.services import competition_service

router = APIRouter(prefix="/api/v1", tags=["competitions"])


@router.post("/competitions", response_model=CompetitionResponse)
def create(req: CreateCompetitionRequest, user: User = Depends(get_current_user),
           db: Session = Depends(get_db)):
    try:
        return competition_service.create_competition(
            db, req.name, req.club_id, req.format,
            req.courts, req.player_ids, req.scheduled_at,
            open_signup=req.open_signup, is_public=req.is_public, max_players=req.max_players,
            signup_deadline=req.signup_deadline,
            creator_user_id=user.id,
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/competitions/open", response_model=list[CompetitionSummary])
def list_open(
    q: str = Query(default=""),
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return competition_service.list_open_competitions(db, user.id, q)


@router.get("/competitions/me", response_model=list[MyCompetitionSummary])
def list_my_competitions(
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return competition_service.list_my_competitions(db, user.id)


@router.post("/competitions/{comp_id}/join", response_model=CompetitionResponse)
def join_open_competition(
    comp_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    try:
        return competition_service.join_open_competition(db, comp_id, user.id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.delete("/competitions/{comp_id}/join", response_model=CompetitionResponse)
def leave_open_competition(
    comp_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    try:
        return competition_service.leave_open_competition(db, comp_id, user.id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/competitions/{comp_id}", response_model=CompetitionResponse)
def get(comp_id: int, user: User = Depends(get_current_user),
        db: Session = Depends(get_db)):
    try:
        return competition_service.get_competition(db, comp_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.patch("/competitions/{comp_id}/start", response_model=CompetitionResponse)
def start(comp_id: int, user: User = Depends(get_current_user),
          db: Session = Depends(get_db)):
    try:
        return competition_service.start_competition(db, comp_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/matches/{match_id}/score", response_model=MatchResponse)
def record_score(match_id: int, req: ScoreRequest,
                 user: User = Depends(get_current_user),
                 db: Session = Depends(get_db)):
    try:
        return competition_service.record_score(db, match_id, req.score_a, req.score_b, user.id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.patch("/competitions/{comp_id}/finish", response_model=CompetitionResponse)
def finish(comp_id: int, user: User = Depends(get_current_user),
           db: Session = Depends(get_db)):
    try:
        return competition_service.finish_competition(db, comp_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
